subprojects/dependency-management/src/main/java/org/gradle/api/internal/artifacts/transform/AttributeMatchingVariantSelector.java

```diff
@@ -161,6 +168,19 @@ private ResolvedArtifactSet doSelect(ResolvedVariantSet producer) {
         return shortestTransforms;
     }
 
+    private List<Pair<ResolvedVariant, ConsumerVariantMatchResult.ConsumerVariant>> disambiguateWithSchema(AttributeMatcher matcher, List<Pair<ResolvedVariant, ConsumerVariantMatchResult.ConsumerVariant>> candidates, ImmutableAttributes componentRequested) {
+        List<AttributeContainerInternal> candidateAttributes = candidates.stream().map(pair -> pair.getRight().attributes).collect(Collectors.toList());
+        List<AttributeContainerInternal> matches = matcher.matches(candidateAttributes, componentRequested);
+        if (matches.size() == 1) {
+            AttributeContainerInternal singleMatch = matches.get(0);
+            return candidates.stream().filter(pair -> pair.getRight().attributes == singleMatch).collect(Collectors.toList());
```

Comment:
I think you should use `.equals` here, otherwise you're relying on an internal implementation detail.

Reply:
I am indeed assuming that this filter-like operation will return the instances sourced from the passed in collection.
It felt like a pretty standard contract to me and gave a nice optimization.
But since I anyway resort to `equals` in the more complicated case, maybe that gain is not worth the risk. Will update after build completes.

Reply:
No, you are assuming that the result of `matches` are the same _instances_ as the ones passed in arguments to the `matches` method. There's no such contract. So basically `==` could fail if `matches` returned say a copy of the initial containers. In practice it will not happen because of our immutable attribute factory, but that's an implementation detail. `equals` will, in any case, to an `==` check in the first place.